### PR TITLE
tiffload: add 16-bit float support

### PR DESCRIPTION
**Pull request by DarthSim**
Hey there 👋

This PR adds support for 16-bit float TIFFs. I didn't use `_Float16` because it's not generally available.

Here are some test images: [tiff3_f16_f32.zip](https://github.com/libvips/libvips/files/12442891/tiff3_f16_f32.zip)

### CODE REVIEW

**Comment by kmilos** [comment]
> I didn't use `_Float16` because it's not generally available.

You could additionally introduce a dependency on [Imath ](https://github.com/AcademySoftwareFoundation/Imath) and let it handle the platform specific/intrinsic stuff if performance is the ultimate goal? And leave this non-Imath fallback if dependency is not wanted/available...

**Comment by DarthSim** [comment]
@kmilos I believe it's a little bit too much for a single rare case 🙂

**Comment by kmilos** [comment]
> Does anyone know if imath has runtime dispatch? Ie. the SSE2 / AVX512 / whatever path is picked at runtime?

I don't think so.

**Comment by jcupitt** [comment]
I had a quick grep and I think it's probably using a look up table by default:

https://github.com/AcademySoftwareFoundation/Imath/blob/main/src/Imath/half.h#L116

You need to build with extra compiler flags to enable the use of intrinsics. Anyway, it's likely to be quicker than bit twiddling, and we probably already have this LUT inside the openexr loader, so it probably makes sense to use it.

File: libvips/foreign/tiff2vips.c [hunk 1]
```diff
@@ -1229,13 +1253,37 @@ rtiff_parse_fourbit(Rtiff *rtiff, VipsImage *out)
 		} \
 	}
 
+/* GREY_LOOP implementation for 16-bit float
+ */
+#define GREY_LOOP_F16 \
+	{ \
+		gshort *p1; \
```
Comment (kmilos) [inline libvips/foreign/tiff2vips.c:1260]: `gushort` everywhere?

**Comment by kmilos** [comment]
> You need to build with extra compiler flags to enable the use of intrinsics.

I guess -march and -mcpu determine if F16C extension is available or not.

> Anyway, it's likely to be quicker than bit twiddling, and we probably already have this LUT inside the openexr loader, so it probably makes sense to use it.

Some idea of difference is already given by the header file:

https://github.com/AcademySoftwareFoundation/Imath/blob/4d4e01c8c9fb9b9dbbefa2e135b21778c9cd331e/src/Imath/half.h#L162-L177

**Comment by kleisauke** [comment]
Once PR #3618 lands, another option would be to use Highway's `PromoteTo()` function, which will map to `_mm{,256,512}_cvtph_ps()` on x86.
https://github.com/google/highway/blob/b39942c49b6e6a24f9dec69b1a02143c8ad21eef/hwy/ops/x86_128-inl.h#L7942-L7945
https://github.com/google/highway/blob/b39942c49b6e6a24f9dec69b1a02143c8ad21eef/hwy/ops/x86_256-inl.h#L6025-L6034
https://github.com/google/highway/blob/b39942c49b6e6a24f9dec69b1a02143c8ad21eef/hwy/ops/x86_512-inl.h#L4624-L4633

This should be safe to use if its availability is verified at both compile- and run time with `HWY_HAVE_FLOAT16` / `hwy::HaveFloat16()`.

I might take a look at this, although 16-bit float TIFFs aren't that common.

File: libvips/foreign/tiff2vips.c (hunk 1 above)
Comment (DarthSim) [reply libvips/foreign/tiff2vips.c]: Oops, thanks. Fixed the typo

**Comment by DarthSim** [comment]
> it's likely to be quicker than bit twiddling

It's definitely quicker. But since 16-bit float TIFF seems to be an extremely rare case, I decided to not go hardcore here.

> we probably already have this LUT inside the openexr loader

It uses `ImfHalfToFloatArray`. We could optionally use it if we have `HAVE_OPENEXR` defined

**Comment by jcupitt** [comment]
Ah I keep forgetting about highway, you're right.

But I agree with @DarthSim --- this is not likely to be a hotspot, so just do something simple for now. We can put in a highway path if this function ever shows up on profiles.

**Comment by DarthSim** [comment]
Applied some code style fixes

**Review by kleisauke — APPROVED** [review]

**Pull request merged**